python/ray/ml/preprocessor.py

```diff
@@ -1,55 +1,104 @@
+                "`fit` cannot be called multiple times. "
+                "Create a new Preprocessor to fit a new Dataset."
+            )
 
-    def transform(self, ds: Dataset) -> Dataset:
-        """Transforms the given dataset.
+        return self._fit(dataset)
 
-        The preprocessor must be fitted first.
+    def _fit(self, dataset: Dataset) -> "Preprocessor":
+        raise NotImplementedError()
 
-        Args:
-            ds (Dataset): Input dataset.
+    def fit_transform(self, dataset: Dataset) -> Dataset:
+        """Fit this Preprocessor to the Dataset and then transform the Dataset.
 
+        Args:
+            dataset (Dataset): Input Dataset.
         Returns:
-            Dataset: Transformed dataset.
+            Dataset: The transformed Dataset.
+        """
+        self.fit(dataset)
+        return self.transform(dataset)
 
+    def transform(self, dataset: Dataset) -> Dataset:
+        """Transform the given dataset.
+
+        Args:
+            dataset (Dataset): Input Dataset.
+        Returns:
+            Dataset: The transformed Dataset.
         """
-        raise NotImplementedError
+        if self._is_fittable and not self.check_is_fitted():
+            raise PreprocessorNotFittedException(
+                "`fit` must be called before `transform`."
+            )
+        return self._transform(dataset)
+
+    def _transform(self, dataset: Dataset) -> Dataset:
+        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
 
     def transform_batch(self, df: DataBatchType) -> DataBatchType:
         """Transform a single batch of data.
 
-        The preprocessor must be fitted first.
-
         Args:
             df (DataBatchType): Input data batch.
-
         Returns:
-            Dataset: Transformed data batch.
+            DataBatchType: The transformed data batch.
         """
-        raise NotImplementedError
 
-    def __getstate__(self) -> dict:
-        return self.__dict__
+        if self._is_fittable and not self.check_is_fitted:
+            raise PreprocessorNotFittedException(
+                "`fit` must be called before `transform_batch`."
+            )
+        return self._transform_batch(df)
+
+    def _transform_batch(self, df: DataBatchType) -> DataBatchType:
+        import pandas as pd
+
+        if not isinstance(df, pd.DataFrame):
+            raise NotImplementedError(
+                "`transform_batch` is currently only implemented for Pandas DataFrames."
+            )
+        return self._transform_pandas(df)
+
+    def _transform_pandas(self, df: "pd.DataFrame") -> "pd.DataFrame":
+        raise NotImplementedError()
+
+    def check_is_fitted(self) -> bool:
+        if hasattr(self, "_is_fitted") and self._is_fitted:
+            return True
 
-    def __setstate__(self, state: dict) -> None:
-        self.__dict__.update(state)
+        fitted_vars = [v for v in vars(self) if v.endswith("_")]
```

Comment:
If you just so `self._is_fitted = True` in the `fit` method of this base class, then would you still need to do this underscore suffixing?

Reply:
Yeah this is what I had originally but changed it to be more sklearn-like (https://github.com/ray-project/ray/pull/22748#discussion_r818042737). cc @Yard1

Reply:
So do we actually need `_is_fitted`? If there are no attributes set in fit (with underscore), then it means no state is persisted after fitting, therefore meaning the preprocessor is not fittable. You can't have a situation where you have only `_is_fitted = True` with an actually fittable preprocessor.

Reply:
Hmmm for my implementation of `Chain` there actually is no fitted state and I set `_is_fitted` to `True` instead. Not sure if there is any state that we would want to store in the `Chain` itself.

Reply:
we can make `Chain`'s `check_is_fitted()` check if all preprocessors inside it are fitted?